### ISSUE TITLE
[QUESTION] How to quantize weight of te.linear in megatron?

### ISSUE THREAD

**Issue by zengwz** [NVIDIA/Megatron-LM#1789]
I want to fake quantize weight of te.linear layer in megatron. If I use pytorch, the code is simple like this:
```
def forward(self, x_quant):
    w_quant = fake_quant(self.weight)
    y = torch.nn.functional.linear(x_quant, w_quant)
    return y
```
However, when applying this method to te.linear, it will raise an error due to `w_quant` lacking `main.grad`. How can I solve this?

**Comment by sbhavani** [comment]
Are you looking to implement custom quantization or debug FP8 training?

If you want FP8 training, we recommend [Transformer Engine](https://github.com/NVIDIA/TransformerEngine) (TE) which is integrated with Megatron.

TE uses `main_grad` for optimized gradient accumulation so manual weight quantization would break this.

**Comment by zengwz** [comment]
I want to implement a manual quantization for BitNet, no FP8 training. Are there any good solutions to achieve a manual quantization without breaking the main grad?

**Comment by sbhavani** [comment]
@zengwz we are working on adding BitNet pre-training support in Megatron. Stay tuned for model updates
